cohttp-lwt/src/s.ml

```diff
@@ -20,23 +20,193 @@ end
+
+    @raise {!exception Connection.Retry} on recoverable errors like the remote
+    endpoint closing
+    the connection gracefully. Even non-idempotent requests are
+    guaranteed to not have been processed by the remote endpoint and
+    should be retried. But beware that a [`Stream] [body] may have been
+    consumed. *)
+type call =
+  ?headers:Http.Header.t -> ?body:Body.t -> Http.Method.t -> Uri.t ->
+  (Cohttp.Response.t * Body.t) Lwt.t
+
+(** The [Connection] module handles a single, possibly pipelined, http
+    connection. *)
+module type Connection = sig
+  module Net : Net
+
+  exception Retry
+  type t
+
+  (** [create ?finalise ?persistent ?ctx endp] connects to [endp]. The
+      connection handle may be used immediately, although the connection
+      may not yet be established.
+      @param finalise called when the connection is closed, but before
+      still waiting requests are failed.
+      @param persistent if [false], a [Connection: close] header is sent and
+      the connection closed as soon as possible. If [true], it is
+      assumed the remote end does support pipelining and multiple
+      requests may be sent even before receiving any reply. By default
+      we wait for the first response to decide whether connection
+      keep-alive and pipelining is suppored.
+      Chunked encoding can only be used when pipelining is supported.
+      Therefore better avoid using chunked encoding on the very
+      first request.
+      @param ctx See [Net.ctx]
+      @param endp The remote address, port and protocol to connect to.
+  *)
+  val create :
+    ?finalise:(t -> unit Net.IO.t) ->
+    ?persistent:bool ->
+    ?ctx:Net.ctx ->
+    Net.endp ->
+    t
+
+  (** Same as [create], but returns d promise which gets fulfilled when
+      the connection is established or rejected when connecting fails. *)
+  val connect :
+    ?finalise:(t -> unit Net.IO.t) ->
+    ?persistent:bool ->
+    ?ctx:Net.ctx ->
+    Net.endp ->
+    t Net.IO.t
+
+  (** Send {e EOF}.
+      On {e TCP} connections send a {e FIN} packet.
+      On {e TLS} connections send a {e close notify}.
+      No new requests can be sent afterwards, but responses may still
+      be received. *)
+  val shutdown : t -> unit
+
+  (** Immediately close connection. All outstanding requests will fail,
+      but non-idempotent requests that already went out on the wire may
+      have produced side-effects. *)
+  val close : t -> unit
+
+  (** If [is_closed connection] is [false] the [connection] still
+      accepts new requests. *)
+  val is_closed : t -> bool
+
+  (** Number of unfulfilled requests. This includes requests already
+      sent out and requests still waitung to be sent. *)
+  val length : t -> int
+
+  (** Request notification on change of [length] and on closing. *)
+  val notify : t -> unit Net.IO.t
+
+  (** Queue a request. Please see {!type:requester}. *)
+  val call : t -> call
+end
+
+module type Connection_cache = sig
+  module IO : IO
```

Comment:
Why do we need to include the `IO` module here?

Reply:
We don't since I changed the type of the handler function to ` ... -> (Response.t, Body.t) Lwt.t` It used to be ` ... -> (Response.t, Body.t) IO.t`.
I removed it.